digdag-core/src/main/java/io/digdag/core/database/AESGCMSecretCrypto.java

```diff
@@ -0,0 +1,177 @@
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.Provider;
+import java.security.SecureRandom;
+import java.security.Security;
+import java.util.Base64;
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static javax.crypto.Cipher.ENCRYPT_MODE;
+
+public class AESGCMSecretCrypto implements SecretCrypto
+{
+    private static final SecureRandom SECURE_RANDOM = new SecureRandom();
+
+    private final SecretKey sharedSecret;
+
+    private static final int AES_KEY_SIZE = 128;
+    private static final int GCM_NONCE_LENGTH = 12;
+    private static final int GCM_TAG_LENGTH = 16;
+
+    private static final int TERM = 1;
+    private static final int VERSION = 1;
+
+    private static final int MAX_PLAINTEXT_LENGTH = 1024;
+
+    private static final int RECORD_SIZE = 2048;
+    private static final int LENGTH_SIZE = 4;
+    private static final int TEXT_SIZE = RECORD_SIZE - LENGTH_SIZE;
+    private static final int TERM_SIZE = 4;
+    private static final int VERSION_SIZE = 4;
+    private static final int OPAQUE_SIZE = TERM_SIZE + VERSION_SIZE + GCM_NONCE_LENGTH + RECORD_SIZE + GCM_TAG_LENGTH;
+
+    public AESGCMSecretCrypto(String sharedSecretBase64)
+    {
+        byte[] sharedSecretRaw = Base64.getDecoder().decode(sharedSecretBase64);
+        Preconditions.checkArgument(sharedSecretRaw.length * 8 == AES_KEY_SIZE);
+        this.sharedSecret = new SecretKeySpec(sharedSecretRaw, "AES");
+    }
+
+    @Override
+    public String encryptSecret(String plainText)
+    {
+        if (plainText.length() > MAX_PLAINTEXT_LENGTH) {
+            throw new IllegalArgumentException("Too long text");
+        }
+
+        byte[] nonce = generateNonce();
+
+        Cipher cipher = cipher(ENCRYPT_MODE, sharedSecret, nonce);
+
+        byte[] plainTextBytes = plainText.getBytes(UTF_8);
+
+        if (plainTextBytes.length > MAX_PLAINTEXT_LENGTH) {
+            throw new IllegalArgumentException("Too long text");
+        }
+
+        byte[] recordBytes = new byte[RECORD_SIZE];
+        ByteBuffer recordBuffer = ByteBuffer.wrap(recordBytes);
+        recordBuffer.putInt(plainTextBytes.length);
+        recordBuffer.put(plainTextBytes);
+
+        byte[] cipherText;
+        try {
+            cipherText = cipher.doFinal(recordBytes);
+        }
+        catch (IllegalBlockSizeException | BadPaddingException e) {
+            throw Throwables.propagate(e);
+        }
+
+        assert cipherText.length == RECORD_SIZE + GCM_TAG_LENGTH;
+
+        byte[] opaque = new byte[OPAQUE_SIZE];
+        ByteBuffer output = ByteBuffer.wrap(opaque);
+
+        output.putInt(TERM);
+        output.putInt(VERSION);
```

Comment:
Having TERM and VERSION is good so that we can change implementation of AESGCMSecretCrypto. But this doesn't solve a problem where we can't change implementation of SecretCrypto at all once we stored some data in DB because the new SecretCrypto can't decode data encrypted by the previous SecretCrypto (e.g. an user started digdag with default settings (DisabledSecretCrypto), and changing settings to use AESGCMSecretCrypto makes already-stored secrets invisible, or makes the server always returning HTTP 500).
I think that better idea is either:

* (A) give up extensibility of SecretCrypto interface completely. DatabaseSecretStore always uses AESGCMSecretCrypto (probably it will be simply renamed to SecretCrypto class) and the class stores encrypted-or-not-flag, algorithm-version, algorithm-name, etc. into the encoded data.

* (B) DatabaseSecretStore stores type of SecretCrypto in addition to the data using new `secrets.crypto_type text` field (text) or `secrets.crypto_type_id int` field (enum). All available SecretCrypto implementations will be injected to DatabaseConfig to decode stored data. In addition to that, default (latest) SecretCrypto is configured to encrypt new records. With this way, AESGCMSecretCrypto maybe doesn't have to include TERM or VERSION because we can create new classes like `AESGCMSecretCryptoV2`, `AESGCMSecretCryptoV3`, ... and let DatabaseSecretStore to choose appropriate one baed on `crypto_type[_id]` field. (TERM is still good to have as a magic header bytes for validation).

I think (B) is better, although it needs more code.